### PR TITLE
WT-12225 Fix RNG generator weakness around mongodb $sample stage

### PR DESCRIPTION
**Pull request by jiechenbo**
The error happens where the records are exactly the same because two cursors are opened close together under one thread. In windows machine clock resolution is low, therefore the numbers ts.tv_nsec and ts.tv_sec are the same. If we only changed the rnd.z seed, we get into a problem like the original code, where rnd.w is the same and rnd.z seed is changing. Therefore the solution used all of the 64 bits from RDTSC.

### CODE REVIEW

**Review by ddanderson — APPROVED** [review]
LGTM

File: test/suite/test_cursor_random03.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            # The random cursor initial seed is time based, reset the initial random seed again
+            # to make sure that we have not generated with the same random numbers
+            cursor = self.session.open_cursor(uri, None, 'next_random=true')
+            found_different = False
+            for i in range(0, 100):
+                self.assertEqual(cursor.next(), 0)
+                current = cursor.get_key()
+                if (random_keys[i] != current):
+                    found_different = True
+                    break
```
[the remark this answers is not in the document]
Comment (jiechenbo) [reply test/suite/test_cursor_random03.py:68]: @kommiharibabu  the bug happened when two of the cursors return back the same exact number of keys. I am happy to just make sure that at least there is one non-match key in this case.

[the remark this answers is not in the document]
Comment (kommiharibabu) [reply test/suite/test_cursor_random03.py:68]: ok, understood. Thanks. Can you add comment to explain the same in the test.

**Review by pmacko86 — APPROVED** [review]
LGTM

File: src/support/rand.c [hunk 2]
```diff
@@ -117,8 +118,17 @@ __wt_random_init_seed(WT_SESSION_IMPL *session, WT_RAND_STATE volatile *rnd_stat
...
+      (uint32_t)ts.tv_sec ^ (uint32_t)WT_LEFT_CIRCULAR_SHIFT32(ts.tv_nsec, 29) ^ DEFAULT_SEED_W;
+    M_Z(rnd) =
+      (uint32_t)ts.tv_nsec ^ (uint32_t)WT_LEFT_CIRCULAR_SHIFT32(ts.tv_sec, 27) ^ DEFAULT_SEED_Z;
+/*
+ * Window's clock does not have a high enough resolution between each tick cycle. Perform an extra
+ * xor against the machine's timestamp counter.
+ */
+#ifdef _WIN32
+    rnd.v ^= __wt_rdtsc();
+#endif
```
Comment (ddanderson) [inline src/support/rand.c:131]: Consider xor-ing `__wt_rdtsc()` unconditionally, and change the comment: `Some system clocks do not have...`. It's a little more future-proof.

**Review by ddanderson — APPROVED** [review]
I made a suggestion for improvement, but I'm fine with this either way.
Nice job tracking this down!

File: src/support/rand.c (hunk 2 above)
Comment (jiechenbo) [reply src/support/rand.c:131]: I would rather not xor __wt_rdtsc() unconditionally yet. Linux systems seem to be doing okay with the current circular shift change, my opinion is that we only start thinking about it if mongodb has gotten failures regarding other machines as well.

**Comment by jiechenbo** [comment]
evergreen retry

**Comment by jiechenbo** [comment]
evergreen retry

**Pull request merged**